pkg/ebpfsyncer/ebpfsyncer.go

```diff
@@ -62,41 +60,105 @@ type ebpfSingleton struct {
 // If isDelete is false then rules will be synchronized for each of the given interfaces.
 func (e *ebpfSingleton) SyncInterfaceIngressRules(
 	ifaceIngressRules map[string][]infv1alpha1.IngressNodeFirewallRules, isDelete bool) error {
+	e.mu.Lock()
+	defer e.mu.Unlock()
 
 	logger := e.log.WithName("syncIngressNodeFirewallResources")
-	logger.Info("Start")
+	logger.Info("Running sync operation", "ifaceIngressRules", ifaceIngressRules, "isDelete", isDelete)
 
 	if e.stats != nil {
 		e.stats.StopPoll()
+		defer func() {
+			if e.c != nil {
+				e.stats.StartPoll(e.c.GetStatisticsMap())
+			}
+		}()
 	}
 
-	for intf, ingress := range ifaceIngressRules {
-		ifMap, err := e.c.IngressNodeFwAttach([]string{intf}, isDelete)
+	// For delete operations, detach all interfaces and run a cleanup, set managed interfaces and the
+	// manager to empty / nil values, then return.
+	// TODO: This should purge all XDP attachments on all interfaces even if we are not managing them.
+	// Alternatively, we need something like method XDPInterfacePurge() that would purse all unmanaged interfaces
+	// both from the rules and from the interface attachements.
+	if isDelete {
+		logger.Info("Running detach and eBPF cleanup operation")
+		for intf := range e.managedInterfaces {
+			_, err := e.c.IngressNodeFwAttach([]string{intf}, true)
+			if err != nil {
+				return err
+			}
+		}
+		e.managedInterfaces = make(map[string]struct{})
+		err := e.c.CleaneBPFObjs()
 		if err != nil {
-			logger.Error(err, "Fail to attach ingress firewall prog")
 			return err
 		}
+		e.c = nil
+
+		return nil
+	}
+
+	// Create a new manager if none exists.
+	var err error
+	if e.c == nil {
+		logger.Info("Creating a new eBPF firewall node controller")
+		e.c, err = nodefwloader.NewIngNodeFwController()
+		if err != nil {
+			return fmt.Errorf("Fail to create nodefw controller instance, err: %q", err)
+		}
+	}
+
+	// Detach any interfaces that were managed by us but that should not be managed any more.
+	// Also delete the rules associated to these interfaces. See TODO below.
+	for intf := range e.managedInterfaces {
+		logger.Info("Running detach operation for interface", "intf", intf)
+		if _, ok := ifaceIngressRules[intf]; !ok {
+			_, err := e.c.IngressNodeFwAttach([]string{intf}, true)
```

Comment:
For future work:
Perhaps split IngressNodeFwAttach into IngressNodeFwAttach and IngressNodeFwDetach.
Then, IngressNodeFwDetach can also automatically purse rules from the ruleset that belong to the interface that's being detached? Or have an extra method for this purse operation.

Reply:
I think merge lib is better place for this